### DISCUSSION TITLE
[Solved] oh-my-bash not working for `su` account

### DISCUSSION THREAD

**Discussion by mhzawadi** [ohmybash/oh-my-bash#581] [General]
Hello 👋🏼 

I have an account that I can only use with `su`, but it seems that oh-my-bash is not working properly. I can see most of it works, but `precmd` and `preexec` stuff is not getting loaded.

How do I make it all work please?

**Comment by mhzawadi** [comment]
FYI: adding `bash-preexec` plugin has worked, now to see what that breaks else where

**Comment by akinomyoga** [comment]
> I can see most of it works, but `precmd` and `preexec` stuff is not getting loaded.

`precmd` and `preexec` are optional features that you need to turn on explicitly. It's not loaded by default intentionally.

Or do you mean there are actual error messages for `precmd` and `preexec`? In my environment, I don't see any errors in loading `precmd` and `preexec`.

> How do I make it all work please?

What do you mean by "all"?

**Reply by mhzawadi** [reply]
So oddly, the `precmd` and `preexec` are both loaded when not `su`ed. so was confused why it doesnt work when `su`ed

**Reply by akinomyoga** [reply]
It depends on the detailed setup, but

- One possibility is that you have a separate setting for the non-su account that loads `bash-preexec`.
- Or another possibility is that you have the `xterm` or `bash-preexec` plugin in `~/.bashrc` of the non-su user account, but you did't have either in `.bashrc` of the su-ed user account. It should be noted that the `xterm` plugin loads `bash-preexec` as a dependency. So if you have the `xterm` plugin in the `plugins` array in `.bashrc`, `bash-preexec` will be automatically loaded even if it is not explicitly specified.

**Reply by mhzawadi** [reply]
I dont have the xterm plugin loaded, just `bashmarks` and `battery`.

But I am using iterm on mac with the iterm helps loaded, so could that tripping me up.

Here is my working config, minus the comments.

```
case $- in
  *i*) ;;
    *) echo "not for you";return;;
esac
export OSH='/Users/matt/.oh-my-bash'
OSH_THEME="bobby-python"
OMB_USE_SUDO=true
completions=(
  composer
  ssh
)
aliases=(
  general
)
plugins=(
  bash-preexec
  bashmarks
  battery
)
source "$OSH"/oh-my-bash.sh
export LESS=eFRX
export HISTCONTROL=$HISTCONTROL${HISTCONTROL+,}ignoredups
if [ -f ~/.bash_aliases ]; then
    . ~/.bash_aliases
fi
source "$HOME/.homesick/repos/homeshick/homeshick.sh"
source "$HOME/.homesick/repos/homeshick/completions/homeshick-completion.bash"
```

**Reply by akinomyoga** [reply]
> But I am using iterm on mac with the iterm helps loaded, so could that tripping me up.

The iTerm2 shell configuration also [contains `bash-preexec.sh`](https://github.com/gnachman/iTerm2/blob/cd9445d59c667cbae57987f1de0100e0a031d666/Resources/shell_integration/iterm2_shell_integration.bash#L86-L439). If it is automatically loaded in the session, the `precmd` and `preexec` features you saw in the non-su user account would probably be provided by iTerm2 (but not by Oh My Bash). I guess the iTerm2 shell integration wouldn't be loaded in the `su`'ed user account, neither do `bash-preexec.sh` embedded in the iTerm2 shell integration.

**Reply by mhzawadi** [reply]
that would do it, I have added the `bash-preexec` plugin and will see how things go.

Thanks 👍🏼
